packages/grafana-data/src/events/EventBus.ts

```diff
@@ -0,0 +1,82 @@
+import EventEmitter from 'eventemitter3';
+import { Unsubscribable, Observable } from 'rxjs';
+import { AppEvent } from './types';
+import { EventBus, LegacyEmitter, BusEventHandler, BusEventType, LegacyEventHandler, BusEvent } from './types';
+
+export class EventBusSrv implements EventBus, LegacyEmitter {
+  private emitter: EventEmitter;
+
+  constructor() {
+    this.emitter = new EventEmitter();
+  }
+
+  $emit<T extends BusEvent>(event: T): void {
+    this.emitter.emit(event.type, event);
+  }
+
+  $on<T extends BusEvent>(typeFilter: BusEventType<T>, handler: BusEventHandler<T>): Unsubscribable {
+    return this.getStream(typeFilter).subscribe({ next: handler });
+  }
+
+  getStream<T extends BusEvent>(eventType: BusEventType<T>): Observable<T> {
+    return new Observable<T>(observer => {
+      const handler = (event: T) => {
+        observer.next(event);
+      };
+
+      this.emitter.on(eventType.type, handler);
+
+      return () => {
+        this.emitter.off(eventType.type, handler);
+      };
+    });
+  }
+
+  /**
+   * Legacy functions
+   */
+  emit<T>(event: AppEvent<T> | string, payload?: T | any): void {
+    // console.log(`Deprecated emitter function used (emit), use $emit`);
+
+    if (typeof event === 'string') {
+      this.emitter.emit(event, { type: event, payload });
+    } else {
+      this.emitter.emit(event.name, { type: event.name, payload });
+    }
+  }
+
+  on<T>(event: AppEvent<T> | string, handler: LegacyEventHandler<T>, scope?: any) {
+    // console.log(`Deprecated emitter function used (on), use $on`);
+
+    // need this wrapper to make old events compatible with old handlers
+    handler.wrapper = (emittedEvent: BusEvent) => {
+      handler(emittedEvent.payload);
+    };
+
+    if (typeof event === 'string') {
+      this.emitter.on(event, handler.wrapper);
+    } else {
+      this.emitter.on(event.name, handler.wrapper);
+    }
+
+    if (scope) {
+      const unbind = scope.$on('$destroy', () => {
+        this.off(event, handler);
+        unbind();
+      });
+    }
+  }
+
+  off<T>(event: AppEvent<T> | string, handler: LegacyEventHandler<T>) {
+    if (typeof event === 'string') {
+      this.emitter.off(event, handler.wrapper);
+      return;
+    }
+
+    this.emitter.off(event.name, handler.wrapper);
+  }
+
+  removeAllListeners() {
```

Comment:
Will we create a singleton of this class? Then this might be a bit dangerous to expose depending on how it is accessible.